src/hotspot/share/logging/logAsyncFlusher.cpp

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+#include "precompiled.hpp"
+#include "jvm.h"
+#include "logging/logAsyncFlusher.hpp"
+#include "logging/logConfiguration.hpp"
+#include "logging/logFileOutput.hpp"
+#include "logging/logHandle.hpp"
+#include "runtime/atomic.hpp"
+
+void AsyncLogMessage::writeback() {
+  if (_message != NULL) {
+    _output.write_blocking(_decorations, _message);
+    os::free(_message);
+  }
+}
+
+void LogAsyncFlusher::enqueue_impl(const AsyncLogMessage& msg) {
+  assert_lock_strong(&_lock);
```

Comment:
Can we enqueue (so, log) from within the async logger itself? If not, add an assert here that this should not be called from the logger thread?

Reply:
enqueue_impl() is private. it's not accessible from outside. 
only public member functions can be seen by clients.

Reply:
> enqueue_impl() is private. it's not accessible from outside.
> only public member functions can be seen by clients.

What I meant is, can logging - hence enqueing - be done from the async thread itself? Can it happen that the logger thread itself logs, would that be a problem, and if yes, should we guard against it?

Reply:
Actually, it could happen. This line is the very first 'log' from AsyncLog Thread. 
```cpp
  void pre_run() override {
    NonJavaThread::pre_run();
    log_debug(logging, thread)("starting AsyncLog Thread tid = " INTX_FORMAT, os::current_thread_id());
  }
```
This logsite is handled exactly same as other logsites. grab the lock, enqueue the cloned message and return -1.